File: sdks/go/multipart_limits.go
package beamnetworksdk

import "fmt"

const (
	multipartMaxPartNumber   = 10_000
	multipartAttemptSlots    = 3
	multipartMaxSourceChunks = multipartMaxPartNumber / multipartAttemptSlots
)

func multipartPartNumber(sourceChunkIndex int, logicalAttemptIndex ...int) (int, error) {
	if sourceChunkIndex < 0 {
		return 0, fmt.Errorf("source_chunk_index must be a non-negative integer")
	}
	attemptIndex := 0
	if len(logicalAttemptIndex) > 0 {
		attemptIndex = logicalAttemptIndex[0]
	}
	if attemptIndex < 0 {
		return 0, fmt.Errorf("logical_attempt_index must be a non-negative integer")
	}
	if sourceChunkIndex >= multipartMaxSourceChunks {
		return 0, fmt.Errorf("source_chunk_index must be less than %d", multipartMaxSourceChunks)
	}
	return (sourceChunkIndex * multipartAttemptSlots) + (attemptIndex % multipartAttemptSlots) + 1, nil
}

File: sdks/typescript/src/multipart-limits.ts
export const MULTIPART_MAX_PART_NUMBER = 10_000;
export const MULTIPART_ATTEMPT_SLOT_COUNT = 3;
export const MULTIPART_MAX_SOURCE_CHUNKS = Math.floor(MULTIPART_MAX_PART_NUMBER / MULTIPART_ATTEMPT_SLOT_COUNT);

export function multipartPartNumber(chunkIndex: number, attemptSlot = 0): number {
  if (!Number.isInteger(chunkIndex) || chunkIndex < 0) {
    throw new Error("source_chunk_index must be a non-negative integer");
  }
  if (chunkIndex >= MULTIPART_MAX_SOURCE_CHUNKS) {
    throw new Error(`source_chunk_index must be less than ${MULTIPART_MAX_SOURCE_CHUNKS}`);
  }
  if (!Number.isInteger(attemptSlot) || attemptSlot < 0 || attemptSlot >= MULTIPART_ATTEMPT_SLOT_COUNT) {
    throw new Error("attempt_slot must be 0, 1, or 2");
  }
  const partNumber = chunkIndex * MULTIPART_ATTEMPT_SLOT_COUNT + attemptSlot + 1;
  return partNumber;
}
